compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2927,6 +2934,26 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             );
         }
     }
+
+    fn suggest_dereferencing_index(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diagnostic,
+        trait_pred: ty::PolyTraitPredicate<'tcx>,
+    ) {
+        if let ObligationCauseCode::ImplDerivedObligation(_) = obligation.cause.code()
+            && self.tcx.is_diagnostic_item(sym::SliceIndex, trait_pred.skip_binder().trait_ref.def_id)
```

Comment:
I think we could look at the 1st (or 2nd?) subst of this `trait_ref` to check if the type is `ty::Slice(..)`. This makes the suggestion [more accurate](https://doc.rust-lang.org/std/slice/trait.SliceIndex.html#impl-SliceIndex%3C%5BT%5D%3E-1), and I think we could bump this to `MachineApplicable` :smiley: 

If this works, then change the wording "consider dereferencing here" -> "dereference this index"